src/consume/mod.rs

```diff
@@ -0,0 +1,202 @@
+/// which this message won't be sent to other consumers. In AWS SQS this is called the [visibility
+/// timeout][AWS], and in GCP PubSub this is the [ack deadline][GCP].
+///
+/// If a message is successfully acknowledged within this time, it will be considered processed and
+/// not delivered to other consumers (and possibly deleted depending on the service's
+/// configuration). A message can conversely be negatively-acknowledged, to indicate e.g.
+/// processing has failed and the message should be delivered again to some consumer. This time
+/// window can also be modified for each message, to allow for longer or shorter message processing
+/// than the default configured time window.
+///
+/// Implementations of this trait do not ack/nack/modify messages themselves, and instead present
+/// this functionality to users with the [`AcknowledgeableMessage`] type. Message processors are
+/// responsible for handling message acknowledgement, including extensions for processing time as
+/// necessary.
+// If we had async drop, sending nacks on drop would be nice. Alas, rust isn't there yet
+///
+/// Bear in mind that message delivery and acknowledgement are all best-effort in distributed
+/// message services. An acknowledged or extended message may still be re-delivered for any number
+/// of reasons, and applications should be made resilient to such events.
+///
+/// [AWS]: https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-visibility-timeout.html
+/// [GCP]: https://cloud.google.com/pubsub/docs/subscriber
+pub trait Consumer {
+    /// The type of acknowledgement tokens produced by the underlying service implementation
+    type AckToken: AcknowledgeToken;
+    /// Errors encountered while streaming messages
+    type Error;
+    /// The stream returned by [`stream`]
+    type Stream: stream::Stream<
+        Item = Result<AcknowledgeableMessage<Self::AckToken, ValidatedMessage>, Self::Error>,
+    >;
+
+    /// Begin pulling messages from the backing message service.
+    ///
+    /// The messages produced by this stream have not been decoded yet. Users should typically call
+    /// [`consume`](Consumer::consume) instead, to produce decoded messages.
+    fn stream(self) -> Self::Stream;
+
+    /// Create a stream of decoded messages from this consumer, using a validator for the given
+    /// [decodable](DecodableMessage) message type.
+    fn consume<M>(self, validator: M::Validator) -> MessageStream<Self::Stream, M::Validator, M>
+        where Self: Sized,
+              M: DecodableMessage,
+    {
+        MessageStream {
+            stream: self.stream(),
+            validator,
+            _message_type: std::marker::PhantomData,
+        }
+    }
+}
+
+/// Messages which can be decoded from a [`ValidatedMessage`] stream.
+pub trait DecodableMessage {
+    /// The error returned when a message fails to decode
+    type Error;
+
+    /// The validator used to decode a validated message
+    type Validator;
```

Comment:
is it a validator or a decoder? whats the difference?

Reply:
the names are all made up. Decoder might be more fitting than Validator here, we could try to set some consistent terminology across the crate. Let me check the python source if they have analogs to this

Reply:
the python uses the same Validator type for encoding and decoding, but its schema checking is slightly different too, which is what motivated the separation I did. I'll name this Decoder and we can follow up later if it matters